portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,56 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    String loginUrl = userService.createLoginURL("/");
+    String logoutUrl = userService.createLogoutURL("/");
+
+    String json = "{";
+    if (userService.isUserLoggedIn()) {
+      json += "\"email\": ";
+      json += "\"" + userService.getCurrentUser().getEmail() + "\"";
+      json += ", ";
+      json += "\"url\": ";
+      json += "\"" + logoutUrl + "\"";
+    } else {
+      json += "\"email\": ";
+      json += "\"null\"";
+      json += ", ";
+      json += "\"url\": ";
+      json += "\"" + loginUrl + "\"";
+    }
+    json += "}";
```

Comment:
I think using string formatting will also help you in this case.

Either that or at least use a [StringBuilder](https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html), as this current solution is unfortunately `O(n**2)` when it could be `O(n)`, because each time you add to a string it creates a new string and handles the work basically on a per-character basis, whereas string builder you input all the strings you want to combine and then call `.toString()` at the end and it handles the per-character level only once at the end.